README.md
docker-mmc-web
==============

[![Docker Pulls](https://img.shields.io/docker/pulls/osixia/mmc-web.svg)][hub]
[![Docker Stars](https://img.shields.io/docker/stars/osixia/mmc-web.svg)][hub]
[![](https://images.microbadger.com/badges/image/osixia/mmc-web.svg)](http://microbadger.com/images/osixia/mmc-web "Get your own image badge on microbadger.com")

[hub]: https://hub.docker.com/r/osixia/mmc-web/

**A docker image to run Mandriva Directory Server**


CHANGELOG.md
# Changelog

## 1.2.0
### Changed
  - Upgrade mmc packages
  - Upgrade baseimage: web-baseimage:1.2.0 (debian buster, php7.3)

## 1.1.1
  - Upgrade mmc packages
  - Optimise apache config
  - Add opcache config
  - Upgrade baseimage: web-baseimage:1.1.1

## 1.1.0
  - Upgrade mmc packages
  - Upgrade baseimage: web-baseimage:1.1.0 (debian stretch, php7)

## 1.0.5
  - Upgrade baseimage: web-baseimage:1.0.0

## 1.0.4
  - Fix root url
  - Upgrade baseimage: web-baseimage:0.1.12

## 1.0.3
  - Upgrade baseimage: web-baseimage:0.1.11
  - Update mmc packages

## 1.0.2
  - Upgrade baseimage: web-baseimage:0.1.10
  - Add MMC_WEB_SERVER_PATH environment variable

## 1.0.1
  - Upgrade baseimage: web-baseimage:0.1.9

## 1.0.0
  - Upgrade mmc packages
  - Upgrade baseimage: web-baseimage:0.1.7
  - Makefile with build no cache

## 0.1.4
  - Upgrade baseimage: web-baseimage:0.1.6

## 0.1.3
  - Upgrade baseimage: web-baseimage:0.1.5
  - Fix group modification
  - Remove listen on http when https is enable

## 0.1.2
  - Upgrade baseimage: web-baseimage:0.1.4
  - Add root url environment variables

## 0.1.1
  - Upgrade baseimage: web-baseimage:0.1.3
  - Rename environment variables

## 0.1.0
  - Initial release


service/mmc/mmc.sh
#!/bin/bash

# -e Exit immediately if a command exits with a non-zero status
if [ ! -e /etc/mmc/docker_bootstrapped ]; then

  ## Apache config
  # Add ssl module
  a2enmod ssl

  # Add rewrite module
  a2enmod rewrite

  #disable default site
  a2dissite 000-default

  # Enable mmc site
  a2ensite mmc

  # Create apache ssl certificate
  /sbin/ssl-create-cert mmc /etc/ssl/certs/mmc-ssl-cert.pem /etc/ssl/private/mmc-ssl-cert.key

  # a ldap container is linked to this phpLDAPadmin container
  if [ -n "${LDAP_NAME}" ]; then
    LDAP_HOST=${LDAP_PORT_389_TCP_ADDR}
    MMC_AGENT_LOGIN=${LDAP_ENV_MMC_AGENT_LOGIN}
    MMC_AGENT_PASSWORD=${LDAP_ENV_MMC_AGENT_PASSWORD}
  else
    LDAP_HOST=${LDAP_HOST}
    MMC_AGENT_LOGIN=${MMC_AGENT_LOGIN}
    MMC_AGENT_PASSWORD=${MMC_AGENT_PASSWORD}
  fi

  # MMC config
  sed -i -e "s?url = https://127.0.0.1:7080?url = https://$LDAP_HOST:7080?" /etc/mmc/mmc.ini
  sed -i -e "s/login = mmc/login = $MMC_AGENT_LOGIN/" /etc/mmc/mmc.ini
  sed -i -e "s/password = s3cr3t/password = $MMC_AGENT_PASSWORD/" /etc/mmc/mmc.ini

  touch /etc/mmc/docker_bootstrapped
fi

source /etc/apache2/envvars
exec apache2 -D FOREGROUND


service/mmc-agent/mmc-agent.sh
#!/bin/sh

## Litle how to: convert ldap schema to ldif

#mkdir -p /etc/ldap/schema/converted
#slaptest -f /etc/ldap/config/convert_to_ldif -F /etc/ldap/schema/converted

#sed -i -e 's/^dn:.*$/dn: cn=mmc,cn=schema,cn=config/; s/^cn:.*$/cn: mmc/; /^structuralObjectClass:.*$/d; /^entryUUID:.*$/d; /^creatorsName:.*$/d; /^createTimestamp:.*$/d; /^entryCSN:.*$/d; /^modifiersName:.*$/d; /^modifyTimestamp:.*$/d' /etc/ldap/schema/converted/cn\=config/cn\=schema/cn=\{4\}mmc.ldif
  
#sed -i -e 's/^dn:.*$/dn: cn=mail,cn=schema,cn=config/; s/^cn:.*$/cn: mail/; /^structuralObjectClass:.*$/d; /^entryUUID:.*$/d; /^creatorsName:.*$/d; /^createTimestamp:.*$/d; /^entryCSN:.*$/d; /^modifiersName:.*$/d; /^modifyTimestamp:.*$/d' /etc/ldap/schema/converted/cn\=config/cn\=schema/cn=\{5\}mail.ldif

# -e Exit immediately if a command exits with a non-zero status
set -e

status () {
  echo "---> ${@}" >&2
}

getBaseDn () {
  IFS="."
  export IFS

  domain=$1
  init=1

  for s in $domain; do
    dc="dc=$s"
    if [ "$init" -eq 1 ]; then
      baseDn=$dc
      init=0
    else
      baseDn="$baseDn,$dc" 
    fi
  done
}

# a ldap container is linked to this phpLDAPadmin container
if [ -n "${LDAP_NAME}" ]; then
  LDAP_HOST=${LDAP_PORT_389_TCP_ADDR}
  
  # Get base dn from ldap domain
  getBaseDn ${LDAP_ENV_LDAP_DOMAIN}

  LDAP_BASE_DN=$baseDn
  LDAP_ADMIN_DN="cn=admin,$baseDn"
else
  LDAP_HOST=${LDAP_HOST}
  LDAP_BASE_DN=${LDAP_BASE_DN}
  LDAP_ADMIN_DN=${LDAP_ADMIN_DN}
fi

if [ ! -e /etc/mmc/agent/docker_bootstrapped ]; then
  status "configuring mmc-agent for first run"

  sed -i -e "s/127.0.0.1/$LDAP_HOST/" /etc/mmc/plugins/base.ini
  sed -i -e "s/dc=mandriva, dc=com/$LDAP_BASE_DN/" /etc/mmc/plugins/base.ini
  sed -i -e "s/password = secret/password = $LDAP_ADMIN_PWD/" /etc/mmc/plugins/base.ini

  mkdir /home/archives

  # Mail plugin

  sed -i -e 's/vDomainSupport = 0/vDomainSupport = 1/g' /etc/mmc/plugins/mail.ini
  sed -i -e 's/vAliasesSupport = 0/vAliasesSupport = 1/g' /etc/mmc/plugins/mail.ini
  cat /etc/mmc/agent/append_to_mail.ini >> /etc/mmc/plugins/mail.ini

  touch /etc/mmc/agent/docker_bootstrapped
else
  status "found already-configured mmc-agent"
fi

exec /usr/sbin/mmc-agent -d


image/service/mmc-web/install.sh
#!/bin/bash -e
# this script is run during the image build

dpkg -i --ignore-depends=libapache2-mod-php5,php5-gd,php5-xmlrpc /container/service/mmc-web/assets/package/mmc-web-base_3.9.90-10_all.deb
dpkg -i /container/service/mmc-web/assets/package/mmc-web-dashboard_3.9.90-10_all.deb
dpkg -i /container/service/mmc-web/assets/package/mmc-web-mail_2.5.1-1_all.deb
dpkg -i /container/service/mmc-web/assets/package/mmc-web-ppolicy_3.9.90-10_all.deb
dpkg -i /container/service/mmc-web/assets/package/mmc-web-sshlpk_2.5.1-1_all.deb

rm -rf /container/service/mmc-agent/assets/package

cat /container/service/mmc-web/assets/php7.3-fpm/pool.conf >> /etc/php/7.3/fpm/pool.d/www.conf
rm /container/service/mmc-web/assets/php7.3-fpm/pool.conf

cp -f /container/service/mmc-web/assets/php7.3-fpm/opcache.ini /etc/php/7.3/fpm/conf.d/opcache.ini
rm /container/service/mmc-web/assets/php7.3-fpm/opcache.ini

# remove apache default host
a2dissite 000-default
rm -rf /var/www/html

# Add apache modules
a2enmod rewrite deflate expires

# change mmc-web default config
sed -i -e "s/#*\s*root\s*=.*/root = \//" /etc/mmc/mmc.ini

# fix d3 lib 
rm /usr/share/mmc/jsframework/d3
ln -sfd /usr/lib/nodejs/d3/build /usr/share/mmc/jsframework/d3
sed -i '1d' /usr/lib/nodejs/d3/build/d3.js

image/service/mmc-web/container-start.sh
#!/bin/bash -e

FIRST_START_DONE="/etc/docker-mmc-web-first-start-done"

# container first start
if [ ! -e "$FIRST_START_DONE" ]; then

  # create mmc-web vhost
  if [ "${MMC_WEB_HTTPS,,}" == "true" ]; then

    # check certificat and key or create it
    /sbin/ssl-helper "/container/service/mmc-web/assets/apache2/certs/$MMC_WEB_HTTPS_CRT_FILENAME" "/container/service/mmc-web/assets/apache2/certs/$MMC_WEB_HTTPS_KEY_FILENAME" --ca-crt=/container/service/mmc-web/assets/apache2/certs/$MMC_WEB_HTTPS_CA_CRT_FILENAME

    # add CA certificat config if CA cert exists
    if [ -e "/container/service/mmc-web/assets/apache2/certs/$MMC_WEB_HTTPS_CA_CRT_FILENAME" ]; then
      sed -i --follow-symlinks "s/#SSLCACertificateFile/SSLCACertificateFile/g" /container/service/mmc-web/assets/apache2/mmc-ssl.conf
    fi

    a2ensite mmc-ssl

  else
    a2ensite mmc
  fi

  # set mmc-agent login and password
  sed -i --follow-symlinks -e "s|#*\s*login\s*=.*|login = \"${MMC_WEB_MMC_AGENT_LOGIN}\"|" /etc/mmc/mmc.ini
  sed -i --follow-symlinks -e "s|#*\s*password\s*=.*|password = \"${MMC_WEB_MMC_AGENT_PASSWORD}\"|" /etc/mmc/mmc.ini

  # set mmc root url
  sed -i --follow-symlinks -e "s|#*\s*root\s*=.*|root = ${MMC_WEB_ROOT_URL}|" /etc/mmc/mmc.ini

  # disable community warning
  sed -i --follow-symlinks -e "s|#*\s*community\s*=.*|community = no|" /etc/mmc/mmc.ini

  # set maxperpage to 50
  sed -i --follow-symlinks -e "s|#*\s*maxperpage\s*=.*|maxperpage = 50|" /etc/mmc/mmc.ini

  # add global keys weakPassword and minsizepassword
  sed -i --follow-symlinks '/\[global\]/a weakPassword = 15'  /etc/mmc/mmc.ini
  sed -i --follow-symlinks '/\[global\]/a minsizepassword = 5'  /etc/mmc/mmc.ini

  # Config servers
  # delete default server config
  sed -i --follow-symlinks '/.*\[server_01\].*/,$d' /etc/mmc/mmc.ini


  server_config() {

    local infos=(${!1})

    for info in "${infos[@]}"
    do
      server_config_value "$info"
    done
  }

  server_config_value() {

    local info_key_value=(${!1})

    local key=${!info_key_value[0]}
    local value=${!info_key_value[1]}

    if [ "$key" = "localcert" ] && [ ! -e "$value" ]; then
      /sbin/ssl-helper "/container/service/mmc-agent-client/assets/certs/mmc-agent-client.tmp.crt" "/container/service/mmc-agent-client/assets/certs/mmc-agent-client.tmp.key" --ca-crt=/container/service/mmc-agent-client/assets/certs/mmc-agent-ca.crt

      # mmc agent need a pem file with the crt and the key
      cat /container/service/mmc-agent-client/assets/certs/mmc-agent-client.tmp.crt /container/service/mmc-agent-client/assets/certs/mmc-agent-client.tmp.key > /container/service/mmc-agent-client/assets/certs/mmc-agent-client.pem
      value="/container/service/mmc-agent-client/assets/certs/mmc-agent-client.pem"
    fi

    if [ "$key" = "cacert" ] && [ ! -e "$value" ]; then
      value="/container/service/mmc-agent-client/assets/certs/mmc-agent-ca.crt"
    fi

    echo "$key = $value" >> /etc/mmc/mmc.ini
  }

  SERVERS=($MMC_WEB_MMC_AGENT_HOSTS)
  i=1
  for server in "${SERVERS[@]}"
  do

    # section var contain a variable name, we access to the variable value and cast it to a table
    infos=(${!server})

    # it's a table of infos
    if [ "${#infos[@]}" -gt "1" ]; then

      echo "[server_$i]" >> /etc/mmc/mmc.ini
      echo "description = ${!infos[0]}" >> /etc/mmc/mmc.ini

      server_config "${infos[1]}"
    fi

    ((i++))
  done

  touch $FIRST_START_DONE
fi

# Fix file permission
chmod 400 /etc/mmc/mmc.ini
chown www-data:www-data /etc/mmc/mmc.ini
chown www-data:www-data -R /usr/share/mmc

exit 0


image/service/mmc-web/startup.sh
#!/bin/bash -e

# set -x (bash debug) if log level is trace
# https://github.com/osixia/docker-light-baseimage/blob/stable/image/tool/log-helper
log-helper level eq trace && set -x


#
# HTTPS config
#
if [ "${MMC_WEB_HTTPS,,}" == "true" ]; then

  log-helper info "Set apache2 https config..."

  # generate a certificate and key if files don't exists
  # https://github.com/osixia/docker-light-baseimage/blob/stable/image/service-available/:ssl-tools/assets/tool/ssl-helper
  ssl-helper ${MMC_WEB_SSL_HELPER_PREFIX} "${CONTAINER_SERVICE_DIR}/mmc-web/assets/apache2/certs/$MMC_WEB_HTTPS_CRT_FILENAME" "${CONTAINER_SERVICE_DIR}/mmc-web/assets/apache2/certs/$MMC_WEB_HTTPS_KEY_FILENAME" "${CONTAINER_SERVICE_DIR}/mmc-web/assets/apache2/certs/$MMC_WEB_HTTPS_CA_CRT_FILENAME"

  # add CA certificat config if CA cert exists
  if [ -e "${CONTAINER_SERVICE_DIR}/mmc-web/assets/apache2/certs/$MMC_WEB_HTTPS_CA_CRT_FILENAME" ]; then
    sed -i "s/#SSLCACertificateFile/SSLCACertificateFile/g" ${CONTAINER_SERVICE_DIR}/mmc-web/assets/apache2/https.conf
  fi

  ln -sf ${CONTAINER_SERVICE_DIR}/mmc-web/assets/apache2/https.conf /etc/apache2/sites-available/mmc-web.conf
#
# HTTP config
#
else
  log-helper info "Set apache2 http config..."
  ln -sf ${CONTAINER_SERVICE_DIR}/mmc-web/assets/apache2/http.conf /etc/apache2/sites-available/mmc-web.conf
fi

#
# Reverse proxy config
#
if [ "${MMC_WEB_TRUST_PROXY_SSL,,}" == "true" ]; then
  echo 'SetEnvIf X-Forwarded-Proto "^https$" HTTPS=on' > /etc/apache2/mods-enabled/remoteip_ssl.conf
fi

a2ensite mmc-web | log-helper debug


FIRST_START_DONE="${CONTAINER_STATE_DIR}/docker-mmc-web-first-start-done"
# container first start
if [ ! -e "$FIRST_START_DONE" ]; then

  # set mmc-agent login and password
  sed -i --follow-symlinks -e "s|#*\s*login\s*=.*|login = \"${MMC_WEB_MMC_AGENT_LOGIN}\"|" /etc/mmc/mmc.ini
  sed -i --follow-symlinks -e "s|#*\s*password\s*=.*|password = \"${MMC_WEB_MMC_AGENT_PASSWORD}\"|" /etc/mmc/mmc.ini

  # set mmc root url
  MMC_WEB_ROOT_URL=$MMC_WEB_SERVER_PATH

  # add trailing slash if needed
  [ ! "${MMC_WEB_ROOT_URL: -1}" == "/" ] && MMC_WEB_ROOT_URL="${MMC_WEB_ROOT_URL}/"
  sed -i --follow-symlinks -e "s|#*\s*root\s*=.*|root = ${MMC_WEB_ROOT_URL}|" /etc/mmc/mmc.ini

  # disable community warning
  sed -i --follow-symlinks -e "s|#*\s*community\s*=.*|community = no|" /etc/mmc/mmc.ini

  # set maxperpage to 50
  sed -i --follow-symlinks -e "s|#*\s*maxperpage\s*=.*|maxperpage = 50|" /etc/mmc/mmc.ini

  # add global keys weakPassword and minsizepassword
  sed -i --follow-symlinks '/\[global\]/a weakPassword = 15'  /etc/mmc/mmc.ini
  sed -i --follow-symlinks '/\[global\]/a minsizepassword = 5'  /etc/mmc/mmc.ini

  # Config servers
  # delete default server config
  sed -i --follow-symlinks '/.*\[server_01\].*/,$d' /etc/mmc/mmc.ini

  # mmc-web agent host config
  host_info(){

    for info in $(complex-bash-env iterate "$1")
    do
      if [ $(complex-bash-env isRow "${!info}") = true ]; then
        local key=$(complex-bash-env getRowKey "${!info}")
        local value=$(complex-bash-env getRowValue "${!info}")

        if [ "$key" = "localcert" ] && [ ! -e "$value" ]; then
          # generate a certificate and key if files don't exists
          # https://github.com/osixia/docker-light-baseimage/blob/stable/image/service-available/:ssl-tools/assets/tool/ssl-helper
          ssl-helper ${MMC_AGENT_SSL_HELPER_PREFIX} "${CONTAINER_SERVICE_DIR}/mmc-agent-client/assets/certs/mmc-agent-client.tmp.crt" "${CONTAINER_SERVICE_DIR}/mmc-agent-client/assets/certs/mmc-agent-client.tmp.key" "${CONTAINER_SERVICE_DIR}/mmc-agent-client/assets/certs/mmc-agent-ca.crt"

          # mmc agent need a pem file with the crt and the key
          cat ${CONTAINER_SERVICE_DIR}/mmc-agent-client/assets/certs/mmc-agent-client.tmp.crt ${CONTAINER_SERVICE_DIR}/mmc-agent-client/assets/certs/mmc-agent-client.tmp.key > ${CONTAINER_SERVICE_DIR}/mmc-agent-client/assets/certs/mmc-agent-client.pem
          value="${CONTAINER_SERVICE_DIR}/mmc-agent-client/assets/certs/mmc-agent-client.pem"
        fi

        if [ "$key" = "cacert" ] && [ ! -e "$value" ]; then
          value="${CONTAINER_SERVICE_DIR}/mmc-agent-client/assets/certs/mmc-agent-ca.crt"
        fi

        echo "$key = $value" >> /etc/mmc/mmc.ini
      fi
    done
  }

  # mmc-web agent config
  i=1
  for host in $(complex-bash-env iterate MMC_WEB_MMC_AGENT_HOSTS)
  do

    if [ $(complex-bash-env isRow "${!host}") = true ]; then

      hostname=$(complex-bash-env getRowKey "${!host}")
      info=$(complex-bash-env getRowValueVarName "${!host}")

      echo "[server_$i]" >> /etc/mmc/mmc.ini
      echo "description = $hostname" >> /etc/mmc/mmc.ini

      host_info "$info"
    else
      exit 1
    fi

    ((i++))
  done

  cp -f /etc/mmc/mmc.ini ${CONTAINER_SERVICE_DIR}/mmc-web/assets/mmc.ini

  touch $FIRST_START_DONE
fi

ln -sf ${CONTAINER_SERVICE_DIR}/mmc-web/assets/mmc.ini /etc/mmc/mmc.ini

# Fix file permission
chmod 400 /etc/mmc/mmc.ini
chown www-data:www-data /etc/mmc/mmc.ini
chown www-data:www-data -R /usr/share/mmc

exit 0
